### PR TITLE
Enc tests

### PR DESCRIPTION
**Pull request by ivanbasov**
**Customer scenario**

These are new unit tests to support C# 7.0 features in ENC.

**Bugs this fixes:**

towards to #17896, #12435 and #12436

**Workarounds, if any**

**Risk**

Low

**Performance impact**

None

**Is this a regression from a previous update?**

**Root cause analysis:**

**How was the bug found?**

Planned

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 1]
```diff
@@ -7183,5 +7183,714 @@ static void F()
...
+            var v1 = CompileAndVerify(compilation1);
+            var md1 = ModuleMetadata.CreateFromImage(v1.EmittedAssemblyData);
+
+            var generation1 = EmitBaseline.CreateInitialBaseline(md1, v1.CreateSymReader().GetEncMethodDebugInfo);
+            var diff2 = compilation1.EmitDifference(
+                generation0,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f1, f2, preserveLocalVariables: true)));
+
+            diff1.VerifyIL("C.G", @"
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >diff1 [](start = 12, length = 5)

diff2

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 2]
```diff
@@ -7183,5 +7183,714 @@ static void F()
...
         }
+
+        [Fact]
+        public void OutVar()
+        {
+            var source0 = MarkedSource(@"
+class C
+{
+    static void F(out int x, out int y) { x = 1; y = 2; }
+    static int G() { F(out int x, out var y); return x + y; }
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >x [](start = 31, length = 1)

You need to mark the declarators that match:
In source0: ```<N:0>x</N:0>``` 
In source1: ```<N:0>x</N:0>``` 
In source2: ```<N:0>x</N:0>```

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 3]
```diff
@@ -7183,5 +7183,714 @@ static void F()
...
+            var diff1 = compilation1.EmitDifference(
+                generation0,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f0, f1, preserveLocalVariables: true)));
+
+            diff1.VerifyIL("C.F", @"
+{
+  // Code size       52 (0x34)
+  .maxstack  1
+  .locals init ([int] V_0,
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >[int] [](start = 16, length = 5)

This is suspicious.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 4]
```diff
@@ -7183,5 +7183,714 @@ static void F()
...
+                bool V_1,
+                [unchanged] V_2,
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: > [unchanged] [](start = 15, length = 12)

This is suspicious.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 5]
```diff
@@ -7183,5 +7183,714 @@ static void F()
...
+            var v1 = CompileAndVerify(compilation1);
+            var md1 = ModuleMetadata.CreateFromImage(v1.EmittedAssemblyData);
+
+            var generation1 = EmitBaseline.CreateInitialBaseline(md1, v1.CreateSymReader().GetEncMethodDebugInfo);
+            var diff2 = compilation1.EmitDifference(
+                generation0,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f1, f2, preserveLocalVariables: true)));
+
+            diff1.VerifyIL("C.F", @"
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >diff1 [](start = 12, length = 5)

diff2

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 6]
```diff
@@ -7183,5 +7183,714 @@ static void F()
...
+
+            var compilation0 = CreateStandardCompilation(source0.Tree, options: ComSafeDebugDll);
+            var compilation1 = compilation0.WithSource(source1.Tree);
+            var compilation2 = compilation1.WithSource(source2.Tree);
+
+            var f0 = compilation0.GetMember<MethodSymbol>("C.F");
+            var f1 = compilation1.GetMember<MethodSymbol>("C.F");
+            var f2 = compilation1.GetMember<MethodSymbol>("C.F");
+
+            var v0 = CompileAndVerify(compilation0);
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >v0 [](start = 16, length = 2)

Add v0.VerifyIL("C.F", ...) to see what the baseline IL is.

**Comment by tmat** [comment]
:clock1:

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 1 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Ouh. Sure! Makes sense

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 5 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Sure. Thanks!

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 6 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Thanks! Done

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 2 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: OK. Done. However, it seems that the produced IL didn't change.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 3 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: I think that this is the example when we change int to bool. Likely, in this case, we should introduce new slots.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 4 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: The same as in another example: it seems that it is reasonable to add new slots here due to the type change.

**Comment by ivanbasov** [comment]
Sent a new iteration. @tmat, could you please review?

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 7]
```diff
@@ -7183,5 +7183,920 @@ static void F()
...
+  IL_0012:  ret
+}
+");
+
+            var md0 = ModuleMetadata.CreateFromImage(v0.EmittedAssemblyData);
+            var generation0 = EmitBaseline.CreateInitialBaseline(md0, v0.CreateSymReader().GetEncMethodDebugInfo);
+            var diff1 = compilation1.EmitDifference(
+                generation0,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f0, f1, preserveLocalVariables: true)));
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Missing argument: GetSyntaxMapFromMarkers(source0, source1)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 8]
```diff
@@ -7183,5 +7183,920 @@ static void F()
...
+}
+");
+            var v1 = CompileAndVerify(compilation1);
+            var md1 = ModuleMetadata.CreateFromImage(v1.EmittedAssemblyData);
+
+            var generation1 = EmitBaseline.CreateInitialBaseline(md1, v1.CreateSymReader().GetEncMethodDebugInfo);
+            var diff2 = compilation2.EmitDifference(
+                generation1,
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f1, f2, preserveLocalVariables: true)));
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Missing argument: GetSyntaxMapFromMarkers(source1, source2)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 7 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Thank you! Fixed in all new 7 tests.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 8 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Thank you! Fixed in all new 7 tests.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 9]
```diff
@@ -7183,5 +7183,920 @@ static void F()
...
+  IL_0011:  ldloc.2
+  IL_0012:  ret
+}
+");
+            var v1 = CompileAndVerify(compilation1);
+            var md1 = ModuleMetadata.CreateFromImage(v1.EmittedAssemblyData);
+
+            var generation1 = EmitBaseline.CreateInitialBaseline(md1, v1.CreateSymReader().GetEncMethodDebugInfo);
+            var diff2 = compilation2.EmitDifference(
+                generation1,
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >generation1 [](start = 16, length = 11)

This is testing the first generation again. We need to test 2nd generation and instead pass in diff1.NextGeneration.

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: ... in all new tests

---
In reply to: [113322020](https://github.com/dotnet/roslyn/pull/18893#discussion_r113322020) [](ancestors = 113322020)

Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Thank you! Fixed

**Comment by ivanbasov** [comment]
need to rebase. will send more

**Comment by ivanbasov** [comment]
I have sent an iteration with diff2 based on diff1 not on the new generation. @tmat, could you please take a look?

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 10]
```diff
@@ -7183,5 +7183,934 @@ static void F()
...
+  IL_000e:  stloc.s    V_6
+  IL_0010:  br.s       IL_0012
+  IL_0012:  ldloc.s    V_6
+  IL_0014:  ret
+}
+");
+        }
+
+        [Fact]
+        public void Tuple_Var()
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >public void Tuple_Var( [](start = 8, length = 22)

I don;t think this test is necessary. It's the same as Tuple_Parenthesized from EnC point of view.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 11]
```diff
@@ -7183,5 +7183,934 @@ static void F()
...
+  IL_0019:  stloc.s    V_5
+  IL_001b:  br.s       IL_001d
+  IL_001d:  ldloc.s    V_5
+  IL_001f:  ret
+}
+");
+        }
+
+        [Fact]
+        public void Tuple_DecompositionIgnoreWithVariableDeclaration()
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: >Tuple_DecompositionIgnoreWithVariableDeclaration [](start = 20, length = 48)

I would also delete this test - it's just adding names to tuple elements. The local types/IL structure is the same as the test above.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 12]
```diff
@@ -3106,5 +3106,330 @@ .maxstack  3
...
+  IL_000c:  stloc.s    V_4
+  IL_000e:  br.s       IL_0010
+ -IL_0010:  ldloc.s    V_4
+  IL_0012:  ret
 }
+", methodToken: diff1.UpdatedMethods.Single());
+        }
+
+        [Fact]
+        public void Tuple_Var()
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: >Tuple_Var [](start = 20, length = 9)

Again, there is no much difference between this one and Tuple_Parenthesized -- they only differ in syntax (var vs explicitly types). What we are testing in EnC tests is emit phase, in which the syntax used to declare a type is irrelevant.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 13]
```diff
@@ -3106,5 +3106,330 @@ .maxstack  3
...
+  IL_002e:  stloc.2
+  IL_002f:  br.s       IL_0031
+ -IL_0031:  ldloc.2
+  IL_0032:  ret
+}
+", methodToken: diff1.UpdatedMethods.Single());
+        }
+
+        [Fact]
+        public void Tuple_DecompositionIgnoreWithVariableDeclaration()
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: >public void Tuple_DecompositionIgnoreWithVariableDeclaration() [](start = 8, length = 62)

Also unnecessary.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 14]
```diff
@@ -3106,5 +3106,330 @@ .maxstack  3
...
+ -IL_0007:  ldloc.0
+  IL_0008:  ldloc.1
+  IL_0009:  add
+  IL_000a:  ldloc.2
+  IL_000b:  add
+  IL_000c:  stloc.s    V_4
+  IL_000e:  br.s       IL_0010
+ -IL_0010:  ldloc.s    V_4
+  IL_0012:  ret
 }
```
Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: We need a couple of pattern matching tests.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 10 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: OK. Removed. Thanks!

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 11 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: OK. Removed. Thanks!

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs (hunk 12 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: OK. Removed. Thanks!

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs (hunk 13 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: OK. Removed. Thanks!

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs (hunk 14 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: Added. Thank you!

**Comment by tmat** [comment]
@jcouv @cston Please review -- should be quick, only adding new tests.

**Comment by cston** [comment]
LGTM

**Comment by ivanbasov** [comment]
Test only. Not Ask mode approval is required. Committing it.

**Pull request merged**